README.md
Midi-Transform
============

Ever wondered how some musical piece would sound if played backwards? What would happen if you would play that piece with your sheet upside down? Midi-Transform lets you make this transformations to MIDI files. 

The transformations available are: Revert, Invert and Change pitch.

[Project page](http://cmrm.github.io/Midi-Transform/)

Download
===========

These are builds with graphical user interface. To use it by command line clone the repository.

* [Windows build](https://github.com/cmrm/midi-transform/releases/download/v1.0.2/miditransform-gui_v1.0.2.exe)
* [Linux build](https://github.com/cmrm/midi-transform/releases/download/v1.0.2/miditransform-gui_v1.0.2.tar.gz)
* [Android APK](https://github.com/cmrm/midi-transform/releases/download/v1.0.2/MidiTransform-1.0.2.apk)

Screenshots
===========
![Windows](extra/qt_screenshot.png?raw=true "Windows") 
![Linux](extra/qt_linux_screenshot.png?raw=true "Linux") 

![Android](extra/android_screenshot.png?raw=true "Android")


Transformations
===============
#### Revert
Revert note order. This transform the MIDI to play backwards.

![Revert example](extra/RevertTransform.png?raw=true "Revert example")

#### Invert
Invert the notes like in a mirror. In MIDI every note has a numerical value. For example, C5, G4, B4 and E5 have numeric values of 72, 67, 71 and 76 respectively. MIDI files can have note values ranging from 0 to 127, so the inversion is done by changing the values to: note_value = 127 - note_value. The new values will be 55, 60, 56 and 51 which correspond to G3, C4, G#3 and D#3. This transformation has the proporty that alters the chord structure from major to minor and viceversa.

![Invert example](extra/InvertTransform.png?raw=true "Invert example")

#### Pitch change
Change pitch of every note. Works by adding or sustracting the pitch change to the note values.

![Pitch change example](extra/PitchTransform.png?raw=true "Pitch change example")

Showcase
===========

Examples of the files produced by the script.

* [Inverse Brahms Symphony No. 3 in F major, Op. 90. Mov. I](https://www.youtube.com/watch?v=kfrJyiVtBUI)
* [Reverse Chopin Ballade No. 1 in G minor, Op. 23](https://www.youtube.com/watch?v=fBDhxt-4A1Y)


Instructions
===========
To use it by command line.
```
Usage: python2 miditransform.py [input file] [output file]
Options:
  -l      , --log                      Write log to files.
  -i      , --invert                   Invert notes on pentagram.
  -r      , --revert                   Revert midi file (play backward).
  -c <arg>, --change_pitch=<arg>       Change pitch of midi file. The argument is the pitch change. E.g: -c 2.
  -t      , --test                     Test similarity of files after transformations.
  -h        --help                     Print usage and exit.
```

A Qt GUI is provided in "miditransform-gui.pyw". It can also be compiled for Android using Kivy and Buildozer.

Requires
===========

* [Python](http://www.python.org/download/) (2.6 or 2.7)
* [PySide](http://qt-project.org/wiki/PySide) For the Qt GUI version.
* [PyGame](http://pygame.org) For the Qt GUI version.
* [Kivy](http://kivy.org) For Android version.

License
===========
Midi-Transform is released into the public domain by the copyright holders.


CHANGELOG.md
v1.0.2
============
* Cut silence at start of reversed track fixed.

v1.0.1
============
* Android version bugfixed.

v1.0.0
============
* Initial release.

main.py
#!/usr/bin/env python
#Encoding: utf-8

#========================================================================================
# title           : main.py
# description     : Kivy GUI for android.
# author          : CMRM.
# usage           : Compile with buildozer.spec.
# python_version  : 2.7
# license         : Public domain.
#========================================================================================

from kivy.uix.floatlayout import FloatLayout
from kivy.uix.boxlayout import BoxLayout
from kivy.uix.button import Button
from kivy.properties import ObjectProperty
from kivy.factory import Factory
from kivy.uix.popup import Popup
from kivy.uix.label import Label
from kivy.app import App
from kivy.logger import Logger
from kivy.config import Config
from kivy.clock import Clock
import os
import tempfile
from miditransform import *

# PyGame support midi on Windows, OSX and Linux.
from sys import platform as _platform
if _platform == "linux" or _platform == "linux2" or _platform == "win32" or _platform == "darwin":
    import pygame
    import pygame.mixer as mixer
    Config.set('graphics', 'width', '500')
    Config.set('graphics', 'height', '500')
    on_android = False
else:
    import android.mixer as mixer
    on_android = True

# Global.
open_filename = ""
file_loaded = False
status_label = 0
temp_dir = ""

transformation_none_gui = 0
transformation_invert_gui = 0
transformation_revert_gui = 0
pitch_switch_gui = 0
pitch_slider_gui = 0

def show_popup(title, text):
    box = BoxLayout(orientation='vertical')
    box.add_widget(Label(text=text))
    button = Button(text='Close', size_hint=(1, 0.3))
    box.add_widget(button)
    popup = Popup(title=title, content=box, auto_dismiss=False, size_hint=(None, None), size=(370, 300))
    button.bind(on_press=popup.dismiss)
    popup.open()

def transform(out_file):
    # Get options.
    if transformation_none_gui.state == "down":
        transformation = "None"
    elif transformation_invert_gui.state == "down":
        transformation = "Invert"
    elif transformation_revert_gui.state == "down":
        transformation = "Revert"
    else:
        transformation = "None"

    if pitch_switch_gui.active:
        change_pitch = True
        pitch_change = int(pitch_slider_gui.value)
    else:
        change_pitch = False

    # Apply transformation to midi file.
    a = MidiFile()
    a.open(open_filename)

    if transformation == "Invert":
        a.invert()
    elif transformation == "Revert":
        a.revert()

    if change_pitch:
        a.change_pitch(pitch_change)

    a.save(out_file)

def set_status(text):
    if status_label != 0:
        status_label.text = text

class LoadDialog(FloatLayout):
    def __init__(self, load, cancel):
        super(LoadDialog, self ).__init__()
        self.load = load
        self.cancel = cancel
        if on_android:
            self.ids.filechooser.rootpath = "/sdcard/"
        else:
            self.ids.filechooser.rootpath = "/home/"


class SaveDialog(FloatLayout):
    def __init__(self, save, cancel):
        super(SaveDialog, self ).__init__()
        self.save = save
        self.text_input = ObjectProperty(None)
        self.cancel = cancel
        if on_android:
            self.ids.filechooser.rootpath = "/sdcard/"
        else:
            self.ids.filechooser.rootpath = "/home/"

class MenuBar(FloatLayout):
    def dismiss_popup(self):
        self._popup.dismiss()

    def show_load(self):
        content = LoadDialog(load=self.load, cancel=self.dismiss_popup)
        self._popup = Popup(title="Load file", content=content, size_hint=(0.9, 0.9))
        self._popup.open()

    def show_save(self):
        global open_filename
        if open_filename:
            content = SaveDialog(save=self.save, cancel=self.dismiss_popup)
            self._popup = Popup(title="Save file", content=content, size_hint=(0.9, 0.9))
            self._popup.open()
        else:
            show_popup("Warning", "No file loaded")

    def about(self):
        show_popup("Info", "MidiTransform\nAuthor: cmrm\nWeb: https://github.com/cmrm/midi-transform")

    def quit(self):
        App.get_running_app().stop()

    def load(self, path, filename):
        global open_filename
        global file_loaded
        try:
            open_filename = os.path.join(path, filename[0]).encode('utf-8')
            self.dismiss_popup()
            if open_filename.endswith('.mid'):
                file_loaded = True
                set_status("File loaded: " + filename[0])
            else:
                open_filename = ""
                show_popup("Warning", "You must select a midi file.")
        except:
            show_popup("Warning", "Invalid file.")

    def save(self, path, filename):
        try:
            save_filename = os.path.join(path, filename)
            self.dismiss_popup()
            if not save_filename.endswith('.mid'):
                save_filename += ".mid"

            transform(save_filename)
            show_popup("Success", "File saved successfully.")
        except:
            show_popup("Warning", "Invalid file.")

class Root(BoxLayout):
    def reset_pitch_slider(self):
        self.ids.pitch_slider.value = 0
        self.ids.sliderval_label.text = str(0)

    def is_playing(self):
        return mixer.music.get_busy()

    def deactivate_gui(self, dt):
        self.ids.play_button.text = "Stop"
        self.ids.options_layout.disabled = True

    def activate_gui(self, dt):
        self.ids.play_button.text = "Play preview"
        self.ids.options_layout.disabled = False

    def start_midi_mixer(self):
        # Start audio mixer.
        freq = 44100
        bitsize = -16
        channels = 2
        buffer = 1024
        mixer.init(freq, bitsize, channels, buffer)

    def check_if_finished(self, dt):
        if not self.is_playing():
            Clock.schedule_once(self.activate_gui)
            Clock.unschedule(self.check_if_finished)

    def start_play(self):
        global open_filename
        temp_file = os.path.join(temp_dir, "temp.mid")
        transform(temp_file)
        try:
            mixer.music.load(temp_file)
            mixer.music.play()
            mixer.music.set_volume(1.0)
            Clock.schedule_once(self.deactivate_gui)
            Clock.schedule_interval(self.check_if_finished, 1 / 30.)
        except:
            Logger.info('Music: Error on play.')

    def stop_play(self):
        self.ids.play_button.text = "Play preview"
        self.ids.options_layout.disabled = False
        mixer.music.stop()

    def play_music(self):
        global file_loaded
        if not self.is_playing():
            if file_loaded:
                Logger.info('Music: Starting to play.')
                self.start_play()
            else:
                show_popup("Warning", "No file loaded")
        else:
            Logger.info('Music: Stopping.')
            self.stop_play()

class MainApp(App):
    def build(self):
        # Config.
        self._app_name = 'MidiTransform'
        global temp_dir
        global on_android

        # Create the app directories on Android.
        if on_android:
            temp_dir = self.user_data_dir
            if not os.path.exists(temp_dir):
                os.makedirs(temp_dir)
        else:
            temp_dir = tempfile.gettempdir()

        # Set the global values for some controllers.
        global status_label
        global transformation_none_gui
        global transformation_invert_gui
        global transformation_revert_gui
        global pitch_switch_gui
        global pitch_slider_gui
        self.icon = "resources/midi-transform-icon.png"
        self.root = Root()
        self.root.start_midi_mixer()
        status_label = self.root.ids.status_label
        transformation_none_gui = self.root.ids.transformation_none
        transformation_invert_gui = self.root.ids.transformation_invert
        transformation_revert_gui = self.root.ids.transformation_revert
        pitch_switch_gui = self.root.ids.pitch_switch
        pitch_slider_gui = self.root.ids.pitch_slider
        return self.root

    def on_stop(self):
        self.root.stop_play()

    def on_pause(self):
        return True

    def on_resume(self):
        pass

Factory.register('Root', cls=Root)
Factory.register('LoadDialog', cls=LoadDialog)
Factory.register('SaveDialog', cls=SaveDialog)

if '__main__' == __name__:
    MainApp().run()

miditransform.py
#!/usr/bin/env python
#Encoding: utf-8

#========================================================================================
# title           : miditransform.py
# description     : Transform midi file.
# author          : CMRM.
# usage           : python miditransform.py [input file] [output file]
# python_version  : 2.7
# license         : Public domain.
#========================================================================================


import midi
import sys
import copy

__version__ = "1.0.2"

class MidiFile:
    def __init__(self):
        self.tracks = []
        self.events = {'track_name': [],
                       'note_on':    [],
                       'note_off':   [],
                       'lyric':      [],
                       'marker':     [],
                       'tempo':      [],
                       'control_change':   [],
                       'program_change':   [],
                       'time_signature':   [],
                       'key_signature' :   [],
                       'smpte_offset'  :   [],
                       'text_meta'     :   []}

    def open(self, read, open_file=True):
        # Read file and organize events in dictionary.
        self.tracks = []
        if open_file:
            track_group = self.pattern = midi.read_midifile(read)
            resolution = track_group.resolution
        else:
            track_group = read

        total_times = []

        for track in track_group:
            bpm = 120
            time = 0

            # Clear dictionary values.
            for v in self.events.keys():
                self.events[v] = []

            # Events will be separeted to work with them individually.
            t = 0
            for event in track:
                if open_file:
                    t += event.tick
                    evt_type = event
                    evt_data = (t, event)
                else:
                    evt_type = event[1]
                    evt_data = event

                if isinstance(evt_type, midi.TrackNameEvent):
                    self.events["track_name"].append(evt_data)
                elif isinstance(evt_type, midi.NoteOnEvent):
                    self.events["note_on"].append(evt_data)
                elif isinstance(evt_type, midi.NoteOffEvent):
                    self.events["note_off"].append(evt_data)
                elif isinstance(evt_type, midi.LyricsEvent):
                    self.events["lyric"].append(evt_data)
                elif isinstance(evt_type, midi.MarkerEvent):
                    self.events["marker"].append(evt_data)
                elif isinstance(evt_type, midi.SetTempoEvent):
                    if open_file:
                        # Calculate elapsed time.
                        time += 60*t / (bpm * resolution)
                        bpm = event.get_bpm()
                    self.events["tempo"].append(evt_data)
                elif isinstance(evt_type, midi.ControlChangeEvent):
                    self.events["control_change"].append(evt_data)
                elif isinstance(evt_type, midi.ProgramChangeEvent):
                    self.events["program_change"].append(evt_data)
                elif isinstance(evt_type, midi.TimeSignatureEvent):
                    self.events["time_signature"].append(evt_data)
                elif isinstance(evt_type, midi.KeySignatureEvent):
                    self.events["key_signature"].append(evt_data)
                elif isinstance(evt_type, midi.SmpteOffsetEvent):
                    self.events["smpte_offset"].append(evt_data)
                elif isinstance(evt_type, midi.TextMetaEvent):
                    self.events["text_meta"].append(evt_data)
            self.tracks.append(self.events.copy())
            if open_file:
                time += 60*t / (bpm * resolution)
                total_times.append(time)
        if open_file:
            self.total_time = max(total_times)

    def save(self, midi_file, log=False):
        pattern = midi.Pattern(resolution=self.pattern.resolution, format=self.pattern.format)
        for track in self.tracks:
            # Put all events on a list.
            events = []
            for i in track["track_name"]:
                events.append(i)
            for i in track["text_meta"]:
                events.append(i)
            for i in track["program_change"]:
                events.append(i)
            for i in track["control_change"]:
                events.append(i)
            for i in track["time_signature"]:
                events.append(i)
            for i in track["key_signature"]:
                events.append(i)
            for i in track["smpte_offset"]:
                events.append(i)
            for i in track["tempo"]:
                events.append(i)
            for i in track["marker"]:
                events.append(i)
            for i in track["lyric"]:
                events.append(i)
            for i in track["note_on"]:
                events.append(i)
            for i in track["note_off"]:
                events.append(i)

            # Sort the list and dump it on the track.
            events.sort(key=lambda x: x[0])
            out_track = midi.Track()

            t_prev = 0
            t = 0
            for i in events:
                t = i[0]
                i[1].tick = t - t_prev
                t_prev = t
                out_track.append(i[1])

            # Write the file.
            eot = midi.EndOfTrackEvent(tick=0)
            out_track.append(eot)
            pattern.append(out_track)
        midi.write_midifile(midi_file, pattern)
        if log:
            open("Log.txt", "w").write(repr(pattern))

    def delay_event(self, track, new_track, event_type, max_time):
        event_times = [evt[0] for evt in track[event_type] if len(evt) != 0]
        if event_times:
            event_max_time = max(event_times)
        else:
            event_max_time = 0

        # Delay event or group of events.
        for i, evt in enumerate(track[event_type]):
            if evt[0] == event_max_time or i == len(track[event_type]) - 1:
                new_track.append((max_time, evt[1]))
            else:
                new_track.append((track[event_type][i+1][0], evt[1]))

    def revert(self, cut_silence = True):
        invertible_controllers = 64, 65, 66, 67, 68, 69
        new_pattern = midi.Pattern(resolution=self.pattern.resolution, format=self.pattern.format)
        new_track_group = []

        # Get the time value of the last event.
        times = [j[0] for track in self.tracks for i in track.keys() for j in track[i] if len(j) != 0]
        if times:
            max_time = max(times)
        else:
            max_time = 0

        for track_num, track in enumerate(self.tracks):
            notes_on = []
            controllers = []
            new_track = []

            # Change order of various events. E.g: 
            # delayevt1 -> noninvertevt1 -> delayevt2 -> noninvertevt2. will be 
            # changed to: noninvertevt1 -> delayevt1 -> noninvertevt2 -> delayevt2
            # so is easier to revert.

            self.delay_event(track, new_track, "track_name", max_time)
            self.delay_event(track, new_track, "tempo", max_time)
            self.delay_event(track, new_track, "time_signature", max_time)
            self.delay_event(track, new_track, "key_signature", max_time)
            self.delay_event(track, new_track, "marker", max_time)
            self.delay_event(track, new_track, "program_change", max_time)
            self.delay_event(track, new_track, "text_meta", max_time)

            # Doc: The Controller Event signals the change in a MIDI channels state.
            # Controller events might be in pairs or just alone. Check both cases.
            # This inverts appeareance order. E.g:
            # noninvertevt1start -> evt1 -> noninvertevt1end -> noninvertevt2start -> evt2 -> noninvertevt2end
            # changes to: noninvertevt1end -> evt1 -> noninvertevt1start -> noninvertevt2end -> evt2 -> noninvertevt2start.
            for i, evt in enumerate(track['control_change']):
                # First check if the controller is invertible (like pedals).
                if evt[1].data[0] in invertible_controllers:
                    controller_found = False
                    # A value in the range 0 - 63 is off, and 64 - 127 is on.
                    if evt[1].data[1] >= 64 and evt[1].data[1] <= 127:
                        controllers.append(evt)
                        tmp = midi.ControlChangeEvent(tick=evt[1].tick, channel=evt[1].channel, data=[evt[1].data[0], 0])
                        new_track.append((evt[0], tmp))
                    else:
                        for j, controller in enumerate(controllers):
                            # Look for controller match.
                            if controller[1].data[0] == evt[1].data[0]:
                                new_track.append((evt[0], controller[1]))
                                controllers.pop(j)
                                controller_found = True
                                break

                            if not controller_found:
                                print("Warning: No matching controller found on invertible controller.")
                                tmp = midi.ControlChangeEvent(tick=evt[1].tick, channel=evt[1].channel, data=[evt[1].data[0], 127])
                                new_track.append((evt[0], tmp))

                # If it's not invertible just append it changing the order.
                else:
                    if i == len(track['control_change']) - 1:
                        new_track.append((max_time, evt[1]))
                    else:
                        new_track.append((track['control_change'][i+1][0], evt[1]))

            if len(controllers) != 0:
                # The unmatched ones are probably at the end of the track.
                print("Warning: Track " + str(track_num) + ", " + str(len(controllers)) + " controller event/s unmatched.")
                for c in controllers:
                    new_track.append((max_time, c))

            # Revert notes (invert appearance order).
            for evt in track['note_on']:
                # Notes on.
                if evt[1].data[1] != 0:
                    notes_on.append(evt)
                    tmp = midi.NoteOnEvent(tick=evt[1].tick, channel=evt[1].channel, data=[evt[1].data[0], 0])
                    new_track.append((evt[0], tmp))

                # Notes off.
                else:
                    note_found = False
                    for i, note in enumerate(notes_on):
                        # Look for note match.
                        if note[1].data[0] == evt[1].data[0]:
                            new_track.append((evt[0], note[1]))
                            notes_on.pop(i)
                            note_found = True
                            break   

                    if not note_found:
                        print("Warning: No matching note found.")
                        tmp = midi.NoteOnEvent(tick=evt[1].tick, channel=evt[1].channel, data=[evt[1].data[0], 100])
                        new_track.append((evt[0], tmp))

            # The same as before.
            for evt in track['note_off']:
                note_found = False
                for i, note in enumerate(notes_on):
                    # Look for note match.
                    if note[1].data[0] == evt[1].data[0]:
                        new_track.append((evt[0], note[1]))
                        notes_on.pop(i)
                        note_found = True
                        break   

                if not note_found:
                    print("Warning: No matching note found.")
                    tmp = midi.NoteOnEvent(tick=evt[1].tick, channel=evt[1].channel, data=[evt[1].data[0], 100])
                    new_track.append((evt[0], tmp))

            if len(notes_on) != 0:
                print("Warning: Track " + str(track_num) + ", " + str(len(notes_on)) + " note/s unmatched.")

            # Lyrics won't be changed.
            for evt in track['lyric']:
                new_track.append(evt)

            # Sort and eliminate the empty time after the reversal.
            new_track.sort(key=lambda x: x[0], reverse=True)
            for i, evt in enumerate(new_track):
                t = max_time - evt[0]
                new_track[i] = (t, evt[1])
            new_track_group.append(list(new_track))

        # Find the time when the first note occurs and cut the silence between.
        if cut_silence:
            noteon_times = [evt[0] for track in new_track_group for evt in track if isinstance(evt[1], midi.NoteOnEvent)]
            if len(noteon_times) > 0:
                min_noteon = min(noteon_times)
                if min_noteon != 0:
                    for i, track in enumerate(new_track_group):
                        for j, evt in enumerate(track):
                            t = evt[0] - min_noteon
                            if t >= 0:
                                new_track_group[i][j] = (t, evt[1])

        self.open(read=new_track_group, open_file=False)

    def change_pitch(self, pitch_var):
        # Add pitch_var to every note value.
        new_pattern = midi.Pattern(resolution=self.pattern.resolution, format=self.pattern.format)
        new_track_group = []
        times = [j[0] for track in self.tracks for i in track.keys() for j in track[i] if len(j) != 0]
        if times:
            max_time = max(times)
        else:
            max_time = 0

        for i, track in enumerate(self.tracks):
            for j, evt in enumerate(track['note_on']):
                tmp = list(evt[1].data)
                tmp[0] += pitch_var
                self.tracks[i]['note_on'][j][1].data = tuple(tmp)

            for j, evt in enumerate(track['note_off']):
                tmp = list(evt[1].data)
                tmp[0] += pitch_var
                self.tracks[i]['note_off'][j][1].data = tuple(tmp)

    def invert(self):
        # Change every note value to 127 - note value.
        new_pattern = midi.Pattern(resolution=self.pattern.resolution, format=self.pattern.format)
        new_track_group = []
        times = [j[0] for track in self.tracks for i in track.keys() for j in track[i] if len(j) != 0]
        if times:
            max_time = max(times)
        else:
            max_time = 0

        for i, track in enumerate(self.tracks):
            for j, evt in enumerate(track['note_on']):
                tmp = list(evt[1].data)
                tmp[0] = 127 - tmp[0]
                self.tracks[i]['note_on'][j][1].data = tuple(tmp)

            for j, evt in enumerate(track['note_off']):
                tmp = list(evt[1].data)
                tmp[0] = 127 - tmp[0]
                self.tracks[i]['note_off'][j][1].data = tuple(tmp)

    def compare(self, midifile):
        # Check how many event coincide perfectly.
        total = 0.0
        match = 0.0
        for i, track in enumerate(self.tracks):
            for key in track.keys():
                for j, event in enumerate(track[key]):
                    try:
                        if event == midifile.tracks[i][key][j]:
                            match += 1.0
                        total += 1.0
                    except:
                        pass

        if total != 0.0:
            percent = 100.0*match/total
            print("Result: " + "%.2f" % percent + " % match.")
        else:
            print("No events to match.")

    def get_length(self):
        # Return total playing time in seconds.
        return self.total_time

    def print_file(self, log_file):
        open(log_file, "w").write(repr(self.pattern))

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("inputfile", help="Input midi file.")
    parser.add_argument("outfile", help="Output midi file.")
    parser.add_argument("-l", "--log", action="store_true", help="Write log to files.")
    parser.add_argument("-r", "--revert", action="store_true", help="Revert midi file.")
    parser.add_argument("-i", "--invert", action="store_true", help="Invert notes on pentagram.")
    parser.add_argument("-c", "--change_pitch", metavar='<pitch_var>', type=int, help="Change pitch of midi file. Argument is pitch change. E.g: -c 2.", default=0)
    parser.add_argument("-t", "--test", action="store_true", help="Test effectivity.")
    args = parser.parse_args()
    log = args.log
    change_pitch = args.change_pitch
    revert = args.revert
    invert = args.invert
    test = args.test

    if change_pitch == 0 and not revert and not invert and not test:
        print("Please, introduce and option. Use -h to see a list of avalable options.")

    a = MidiFile()
    a.open(args.inputfile)
    if log:
        a.print_file("Original.txt")

    if revert:
        a.revert()
    if invert:
        a.invert()
    if change_pitch != 0:
        a.change_pitch(change_pitch)
    if test:
        # Test revert.
        print("Test: Revert.")
        orig = copy.copy(a)
        a.revert(cut_silence = False)
        a.revert(cut_silence = False)
        orig.compare(a)

        # Test invert.
        print("Test: Invert.")
        orig = copy.copy(a)
        a.invert()
        a.invert()
        orig.compare(a)

        # Test change pitch.
        print("Test: Change pitch.")
        orig = copy.copy(a)
        a.change_pitch(1)
        a.change_pitch(-1)
        orig.compare(a)

    a.save(args.outfile, log)
    print("Done.")